app/flatpak-cli-transaction.c

```diff
@@ -514,144 +619,187 @@ print_permissions (FlatpakTablePrinter *printer,
+          self->uninstalling = TRUE;
+          break;
+        case FLATPAK_TRANSACTION_OPERATION_INSTALL:
+        case FLATPAK_TRANSACTION_OPERATION_INSTALL_BUNDLE:
+          self->installing = TRUE;
+          break;
+        case FLATPAK_TRANSACTION_OPERATION_UPDATE:
+          self->updating = TRUE;
+          break;
+        default:;
+        }
+    }
+
+  printer = self->printer = flatpak_table_printer_new ();
+  i = 0;
+  flatpak_table_printer_set_column_title (printer, i++, _("   "));
+  flatpak_table_printer_set_column_title (printer, i++, _("ID"));
+  flatpak_table_printer_set_column_title (printer, i++, _("Arch"));
+  flatpak_table_printer_set_column_title (printer, i++, _("Version"));
+
+  if (self->installing + self->updating + self->uninstalling > 1)
+    flatpak_table_printer_set_column_title (printer, i++, _("Change"));
 
-      if (!found_one)
-        g_print (_("Uninstalling from %s:\n"), self->name);
-      found_one = TRUE;
-      g_print ("%s\n", pref);
+  if (self->installing || self->updating)
+    {
+      flatpak_table_printer_set_column_title (printer, i++, _("Remote"));
+      flatpak_table_printer_set_column_title (printer, i++, _("Download"));
     }
 
-  found_one = FALSE;
   for (l = ops; l != NULL; l = l->next)
     {
       FlatpakTransactionOperation *op = l->data;
       FlatpakTransactionOperationType type = flatpak_transaction_operation_get_operation_type (op);
       const char *ref = flatpak_transaction_operation_get_ref (op);
       const char *remote = flatpak_transaction_operation_get_remote (op);
-      const char *commit = flatpak_transaction_operation_get_commit (op);
-      GKeyFile *metadata = flatpak_transaction_operation_get_metadata (op);
-      const char *pref = strchr (ref, '/') + 1;
+      g_auto(GStrv) parts = flatpak_decompose_ref (ref, NULL);
 
-      if (type != FLATPAK_TRANSACTION_OPERATION_INSTALL &&
-          type != FLATPAK_TRANSACTION_OPERATION_INSTALL_BUNDLE)
-        continue;
+      flatpak_table_printer_add_column (printer, "   ");
+      flatpak_table_printer_add_column (printer, parts[1]);
+      flatpak_table_printer_add_column (printer, parts[2]);
+      flatpak_table_printer_add_column (printer, parts[3]);
+
+      if (self->installing + self->updating + self->uninstalling > 1)
+        flatpak_table_printer_add_column (printer, op_shorthand[type]);
 
-      if (!found_one)
+      if (type == FLATPAK_TRANSACTION_OPERATION_INSTALL ||
+          type == FLATPAK_TRANSACTION_OPERATION_INSTALL_BUNDLE ||
+          type == FLATPAK_TRANSACTION_OPERATION_UPDATE)
         {
-          g_print (_("Installing in %s:\n"), self->name);
-          printer = flatpak_table_printer_new ();
-          found_one = TRUE;
+          g_autoptr(FlatpakRef) rref = flatpak_ref_parse (ref, NULL);
+          guint64 installed_size;
+          guint64 download_size;
+          g_autofree char *formatted = NULL;
+
+          flatpak_installation_fetch_remote_size_sync (inst,
```

Comment:
This call is not ideal. It will be doing a bunch of i/o getting the summary, depending on cache hits, p2p status and whatnot. The idea of the transaction is to do all the i/o up-front, so if we need the size we should have gotten that up-front when resolving the refs to the transaction.

Reply:
Also, I wonder if showing the download size here really makes anyone happy, because the download size is the maximum download size (from-scratch), not the actual size we will download for an update.

Reply:
so... add a transaction flag ?

Reply:
I've added some transaction api for this